components/wifi/esp8266-driver/ESP8266Interface.cpp

```diff
@@ -502,15 +505,12 @@ int ESP8266Interface::socket_send(void *handle, const void *data, unsigned size)
         return NSAPI_ERROR_NO_SOCKET;
     }
 
-    unsigned long int sendStartTime = rtos::Kernel::get_ms_count();
-    do {
-        status = _esp.send(socket->id, data, size);
-    } while ((sendStartTime - rtos::Kernel::get_ms_count() < 50)
-            && (status != NSAPI_ERROR_OK));
+    status = _esp.send(socket->id, data, size);
 
-    if (status == NSAPI_ERROR_WOULD_BLOCK) {
-        debug("Enqueuing the event call");
-        _global_event_queue->call_in(100, callback(this, &ESP8266Interface::event));
+    if (status == NSAPI_ERROR_WOULD_BLOCK && !_cbs[socket->id].deferred) {
+        tr_debug("Postponing SIGIO from the device");
+        _cbs[socket->id].deferred = true;
+        _global_event_queue->call_in(100, callback(this, &ESP8266Interface::event_deferred));
```

Comment:
Can this be larger? We already saw that in `busy p` the 100ms is not enough for device to recover.